### PR TITLE
Add embedding based LAION retrieval component

### PR DESCRIPTION
**Pull request by ChristiaensBert**
Draft PR for clip retrieval based on embeddings

Dask DataFrame Structure:
                   id  source
npartitions=1                
10_laion       string  string
9_laion           ...     ...
Dask Name: getitem, 28 graph layers
{'source': 'string', 'id': 'string'}
Traceback (most recent call last):
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py", line 98, in <module>
    component.run()
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/fondant/component.py", line 94, in run
    self._write_data(dataframe=output_df, manifest=output_manifest)
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/fondant/component.py", line 83, in _write_data
    data_writer.write_index(dataframe)
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/fondant/data_io.py", line 143, in write_index
    upload_index_task = self._create_write_dataframe_task(
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/fondant/data_io.py", line 115, in _create_write_dataframe_task
    write_task = dd.to_parquet(
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/venv/lib/python3.9/site-packages/dask/dataframe/io/parquet/core.py", line 963, in to_parquet
    i_offset, fmd, metadata_file_exists, extra_write_kwargs = engine.initialize_write(
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/venv/lib/python3.9/site-packages/dask/dataframe/io/parquet/arrow.py", line 710, in initialize_write
    inferred_schema = pyarrow_schema_dispatch(
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/venv/lib/python3.9/site-packages/dask/utils.py", line 642, in __call__
    return meth(arg, *args, **kwargs)
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/venv/lib/python3.9/site-packages/dask/dataframe/backends.py", line 213, in get_pyarrow_schema_pandas
    return pa.Schema.from_pandas(obj)
  File "pyarrow/types.pxi", line 1492, in pyarrow.lib.Schema.from_pandas
  File "/Users/bertchristiaens/Bitbucket/fondant_ml6/fondant/venv/lib/python3.9/site-packages/pyarrow/pandas_compat.py", line 530, in dataframe_to_types
    type_ = pa.array(c.head(0), from_pandas=True).type
AttributeError: 'Index' object has no attribute 'head'

### CODE REVIEW

File: fondant/data_io.py
Comment (RobbeSneyders) [inline fondant/data_io.py:1]: Print statements in this file can be removed.

File: examples/pipelines/finetune_stable_diffusion/components/load_from_hub/src/main.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ def load(self, *, dataset_name: str) -> dd.DataFrame:
         dask_df = dask_df.astype({"id": "string"})
         dask_df["source"] = "hub"
 
+        dask_df = dask_df.sample(frac=0.05)
```
Comment (RobbeSneyders) [inline examples/pipelines/finetune_stable_diffusion/components/load_from_hub/src/main.py:53]: Can be removed I think

File: examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        # add id and source columns
+        dataframe["id"] = dataframe.assign(id=1).id.cumsum()
+        dataframe["source"] = "laion"
+        dataframe = dataframe.astype({'id': 'string', 'source': 'string', 'images_url': 'string'})
+
+        # reorder columns
+        dataframe = dataframe[["id", "source", "images_url"]]
+        dataframe = dataframe.reset_index(drop=True)
+        print(type(dataframe))
```
Comment (RobbeSneyders) [inline examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py:88]: Can be removed

File: examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        """
+        client = ClipClient(
+            url="https://knn.laion.ai/knn-service",
+            indice_name="laion5B-L-14",
+            num_images=num_images,
+            aesthetic_score=aesthetic_score,
+            aesthetic_weight=aesthetic_weight,
+            modality=Modality.IMAGE,
+        )
+        dataframe = dataframe.sample(frac=0.2)
```
Comment (RobbeSneyders) [inline examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py:68]: Can be removed.

File: examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        dataframe = dataframe.astype({'id': 'string', 'source': 'string', 'images_url': 'string'})
+
+        # reorder columns
+        dataframe = dataframe[["id", "source", "images_url"]]
+        dataframe = dataframe.reset_index(drop=True)
+        print(type(dataframe))
+        # uid index is id+"_"+source
+        dataframe["uid"] = dataframe["id"] + "_" + dataframe["source"]
+        # set as index
+        dataframe = dataframe.set_index("uid")
```
Comment (RobbeSneyders) [inline examples/pipelines/finetune_stable_diffusion/components/embedding_based_laion_retrieval/src/main.py:92]: This is handled by the component class

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @ChristiaensBert!

File: fondant/data_io.py
Comment (ChristiaensBert) [reply fondant/data_io.py:1]: Yes, so this is not a final PR. It was made so George could look at it, since there is some weird stuff going on.

**Pull request closed without merging**